nova/virt/vmwareapi/vmops.py

```diff
@@ -3125,3 +3129,91 @@ def update_vmref_cache(self):
                 continue
 
             vm_util.vm_ref_cache_update(vm_uuid, props['obj'])
+
+    def sync_server_group(self, context, sg_uuid):
+        LOG.debug('Starting sync for server-group %s', sg_uuid)
+
+        @utils.synchronized('vmware-server-group-{}'.format(sg_uuid))
+        def _sync_sync_server_group(self, context, sg_uuid):
+            rule_prefix = '{}{}'.format(constants.DRS_PREFIX, sg_uuid)
+
+            # retrieve the server-group with its members
+            try:
+                sg = objects.instance_group.InstanceGroup.get_by_uuid(context,
+                                                                      sg_uuid)
+            except exception.InstanceGroupNotFound:
+                LOG.info('Server-group %s cannot be found in DB', sg_uuid)
+                # check if we have it in the vCenter. if yes, it's an orphan
+                # and needs deletion
+                rules = cluster_util.get_rules_by_prefix(
+                    self._session, self._cluster_ref, rule_prefix)
+                for rule in rules:
+                    LOG.debug('Deleting DRS rule %s as orphan', rule.name)
+                    cluster_util.delete_vm_rule(
+                        self._session, self._cluster_ref, rule)
+                    LOG.info('Deleted rule %s as orphan', rule.name)
+                LOG.debug('Sync for server-group %s done', sg_uuid)
+                return
+
+            # retrieve the instances, because sg.members contains all members
+            # and we need to filter them for our host
+            InstanceList = objects.instance.InstanceList
+            instances = InstanceList.get_by_host(context, self._compute_host,
+                                                 expected_attrs=[])
+            instances_by_uuid = {i.uuid: i for i in instances}
+            instance_uuids = set(instances_by_uuid)
+            member_uuids_on_host = set(sg.members) & instance_uuids
```

Comment:
This is my main problem with this approach. We now have to get list of all instances for each update, which is worse than doing it once and syncing everything in maybe three seconds.

Couldn't we move the filtering to the db this way:
```python
filters = {"host": , "uuid": sg.members, "deleted": False}
instances = InstanceList.get_by_filters(context, filters, expected_attrs=[])
instances_by_uuid = {i.uuid: i for i in instances}
member_uuids_on_host = instances_by_uuid.keys()
```

Reply:
Yes, you're right, we can turn this query around now and filter out the members belonging to the host. :+1: